sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java

```diff
@@ -0,0 +1,214 @@
+ * new SimpleFunction<Iterable<String>, Iterable<String>>() {
+ *   {@literal @}Override
+ *   public Iterable<String> apply(Iterable<String> input) {
+ *     ArrayList<String> results = callWebService(input);
+ *     return results;
+ *     }
+ *  };
+ *  ...
+ *  Pipeline pipeline = Pipeline.create(...);
+ *  ...
+ *  long batchSize = 100L;
+ *  pipeline.apply(BatchingParDo.via(batchSize, perBatchFn))
+ *          .setCoder(StringUtf8Coder.of());
+ *  pipeline.run();
+ * }</pre>
+ * *
+ */
+public class BatchingParDo<K, InputT, OutputT>
+    extends PTransform<PCollection<KV<K, InputT>>, PCollection<KV<K, OutputT>>> {
+
+  private final long batchSize;
+  private final SimpleFunction<? super Iterable<InputT>, ? extends Iterable<OutputT>> perBatchFn;
+
+  private BatchingParDo(
+      long batchSize,
+      SimpleFunction<? super Iterable<InputT>, ? extends Iterable<OutputT>> perBatchFn) {
+    this.batchSize = batchSize;
+    this.perBatchFn = perBatchFn;
+  }
+
+  public static <K, InputT, OutputT> BatchingParDo<K, InputT, OutputT> via(
+      long batchSize,
+      SimpleFunction<? super Iterable<InputT>, ? extends Iterable<OutputT>> perBatchFn) {
+    return new BatchingParDo<>(batchSize, perBatchFn);
+  }
+
+  @Override
+  public PCollection<KV<K, OutputT>> expand(PCollection<KV<K, InputT>> input) {
+    Duration allowedLateness = input.getWindowingStrategy().getAllowedLateness();
+
+    PCollection<KV<K, OutputT>> output =
+        input.apply(
+            ParDo.of(
+                new BatchingDoFn<>(
+                    batchSize,
+                    perBatchFn,
+                    allowedLateness,
+                    (Coder<InputT>) input.getCoder().getCoderArguments().get(1),
+                    (Coder<K>) input.getCoder().getCoderArguments().get(0))));
+    return output;
+  }
+
+  @VisibleForTesting
+  static class BatchingDoFn<K, InputT, OutputT> extends DoFn<KV<K, InputT>, KV<K, OutputT>> {
+
+    private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(BatchingDoFn.class);
+    private static final String END_OF_WINDOW_ID = "endOFWindow";
+    private static final String BATCH_ID = "batch";
+    private static final String NUM_ELEMENTS_IN_BATCH_ID = "numElementsInBatch";
+    private static final String TIMER_ALREADY_SET_ID = "timerAlreadySet";
+    private static final String KEY_ID = "key";
+    private final long batchSize;
+    private final SimpleFunction<? super Iterable<InputT>, ? extends Iterable<OutputT>> perBatchFn;
+    private final Duration allowedLateness;
+
+    @TimerId(END_OF_WINDOW_ID)
+    private final TimerSpec timer = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+    @StateId(TIMER_ALREADY_SET_ID)
+    private final StateSpec<Object, ValueState<Integer>> timerAlreadySetForWindow;
+
+    @StateId(BATCH_ID)
+    private final StateSpec<Object, BagState<InputT>> batchSpec;
+
+    @StateId(NUM_ELEMENTS_IN_BATCH_ID)
+    private final StateSpec<Object, ValueState<Long>> numElementsInBatchSpec;
```

Comment:
This can be a `combiningValue` state, which I believe more accurately models the way in which you actually use the state - you add to the state, and then read-back to see if you're done with a batch.